### ISSUE TITLE
[@typescript-eslint/no-unnecessary-condition] Nullable object can still fail the linter

### ISSUE THREAD

**Issue by astorije** [typescript-eslint/typescript-eslint#1897]
**Repro**

```yaml
rules:
  '@typescript-eslint/no-unnecessary-condition':
    - error
    - ignoreRhs: true
      checkArrayPredicates: true
```

```tsx
import React, { FunctionComponent } from 'react';

export const Foo: FunctionComponent<{ opt?: string; req: string }> = props => (
  <>{props.children || props.req || props.opt}</>
);

export const Bar: FunctionComponent<{ opt?: string; req: string }> = props => (
  <>{props.req || props.opt}</>
);

export const Baz: FunctionComponent<{ req: string }> = props => (
  <>{props.children || props.req}</>
);
```

**Expected Result**

Per [the `@types/react` types](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/react/index.d.ts#L825), `children` is optional:

```tsx
type PropsWithChildren<P> = P & { children?: ReactNode };
```

Because of this, I would expect all examples above to pass the rule.

**Actual Result**

For a reason I do not understand, the first example (`Foo`) fails the linter:

```tsx
     Unnecessary conditional, value is always truthy

     src/Foobar.tsx:4:6
     2 |
     3 | export const Foo: FunctionComponent<{ opt?: string; req: string }> = props => (
   > 4 |   <>{props.children || props.req || props.opt}</>
       |      ^
     5 | );
```

But we know from the types that `props.children` is _not_ always truthy.

Additionally:

- If the issue is indeed that `props.children` is always truthy, `Baz` should fail
- I was wondering if the issue was because `props.req` is non-nullable (but `''` is still falsy) but `Bar` doesn't fail either

I can't seem to find another combination that makes the linter scream, unless I'm missing something.

**Additional Info**

Apologies if this has already been reported. I went through existing tickets, opened and closed, and didn't find something similar.

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `2.27.0` |
| `@typescript-eslint/parser`        | `2.27.0` |
| `TypeScript`                       | `3.8.3` |
| `ESLint`                           | `6.8.0` |
| `node`                             | `12.14.0` |
| `yarn`                              | `1.22.4` |

**Comment by bradzacher** [comment]
I'm unable to repro this against master.

Do you have `strictNullChecks` turned on?

**Comment by astorije** [comment]
I do have it turned on yeah.

I'm not testing against master though, but 2.27. I could give it a shot against 2.28, but haven't looked at the changelog to see if there was a relevant change yet.

**Comment by bradzacher** [comment]
would you be able to chuck it into a repro repo?
there might be some specific combination of `@types/react` / typescript that triggers the bug (eg #1868 requires a specific version of the types package)

**Comment by astorije** [comment]
Hey @bradzacher, sorry for the lack of response! It seems like the issue disappeared while upgrading to v3, so it's probably something that got fixed after v2.27.0. I'll re-open if it reappears.
Thanks again for your amazing work!!
